src/solenoid/controls.clj

```diff
@@ -16,10 +16,21 @@
     ;; when controllers are removed, remove any watches associated with affected cursor paths
     (let [removed (set/difference (set (keys old)) (set (keys new)))]
       (when (seq removed)
-        (let [watches (u/get-watches registry)
-              paths   (mapcat #(filter (fn [k]
-                                         (and (vector? k)
-                                              (= (first k) %))) watches) removed)]
+        (let [cb-watches (->> (keys new)
+                              (filter #(str/starts-with? (name %) "controlblock"))
+                              (mapcat #(u/get-watches (get-in new [% :state])))
+                              (filter #(removed (last %))))
+              watches    (u/get-watches registry)
+              paths      (mapcat #(filter (fn [k]
+                                            (and (vector? k)
+                                                 (= (first k) %))) watches) removed)]
+          ;; any removed control blocks might have been dependents of a still-existing block.
+          ;; In such cases, old watches on still-existing blocks must be removed so that we don't get any NPEs
+          (when (seq cb-watches)
+            (doseq [cb (keys new)
+                    w cb-watches]
+              (swap! registry update-in [cb :state] (fn [ref]
+                                                      (when ref (remove-watch ref w))))))
```

Comment:
This addition is critical to allowing dependent controlblocks because it cleans up watches when things are removed from the registry. Without this, after a user deletes a controlblock that was linked to a still-existing block, the server would try to update/send the value for that controlblock that was deleted, since the watch was still attached.

Now, as blocks are removed, so are the watches that pointed to them.